### PR TITLE
Acctest service

### PR DESCRIPTION
**Pull request by TheMattSchiller**
This pull request completes the service resource, and acceptance tests the lifecycle of the service resource

### CODE REVIEW

File: banyan/resource_service.go [hunk 1]
```diff
@@ -229,36 +195,29 @@ func resourceService() *schema.Resource {
...
 												"cidr": {
 													Type:         schema.TypeString,
-													Required:     true,
+													Optional:     true,
 													ValidateFunc: validateCIDR(),
 												},
 												"port": {
-													Type:         schema.TypeInt,
-													Required:     true,
-													ValidateFunc: validatePort(),
```
Comment (dd76) [inline banyan/resource_service.go:245]: ValidatePort() Any reason this was removed.

File: banyan/resource_service.go [hunk 2]
```diff
@@ -866,25 +833,18 @@ func resourceServiceCreate(ctx context.Context, d *schema.ResourceData, m interf
...
+			hts, ok := clientCIDRItemMap["host_tag_selector"].([]interface{})
 			if !ok {
-				diagnostics = diag.Errorf("Couldn't type assert host_tag_selectors")
+				diagnostics = diag.Errorf("couldn't type assert host_tag_selector with type: %v", reflect.TypeOf(clientCIDRItemMap["host_tag_selector"]))
 				return
 			}
-			for _, hostTagSelectorItem := range hostTagSelectors {
-				hostTagSelectorItemMap, ok := hostTagSelectorItem.(map[string]interface{})
-				if !ok {
-					diagnostics = diag.Errorf("Couldn't type assert host tag selector item map, actually has type of: %v", reflect.TypeOf(hostTagSelectorItem))
```
Comment (dd76) [inline banyan/resource_service.go:877]: If err != nil, return is missing

File: banyan/resource_service.go [hunk 3]
```diff
@@ -946,52 +906,39 @@ func resourceServiceCreate(ctx context.Context, d *schema.ResourceData, m interf
...
 			if !ok {
-				diagnostics = diag.Errorf("Couldn't type assert frontend_address")
+				diagnostics = diag.Errorf("Couldn't type assert frontend_address, has type %v", reflect.TypeOf(jj["frontend_address"]))
 				return
 			}
 			for _, frontEndAddressItem := range frontEndAddress {
+				newFrontEndAddress := service.FrontendAddress{}
 				frontEndAddressItemMap, ok := frontEndAddressItem.(map[string]interface{})
 				if !ok {
-					diagnostics = diag.Errorf("Couldn't type assert frontend_address item value %+v", reflect.TypeOf(frontEndAddressItem))
```
Comment (dd76) [inline banyan/resource_service.go:955]: Again returns are removed, why?

File: banyan/resource_service.go [hunk 4]
```diff
@@ -946,52 +906,39 @@ func resourceServiceCreate(ctx context.Context, d *schema.ResourceData, m interf
 
 			frontEndAddress, ok := jj["frontend_address"].([]interface{})
 			if !ok {
-				diagnostics = diag.Errorf("Couldn't type assert frontend_address")
+				diagnostics = diag.Errorf("Couldn't type assert frontend_address, has type %v", reflect.TypeOf(jj["frontend_address"]))
 				return
 			}
 			for _, frontEndAddressItem := range frontEndAddress {
+				newFrontEndAddress := service.FrontendAddress{}
```
Comment (dd76) [inline banyan/resource_service.go:913]: Why these changes? previous one works.

File: banyan/resource_service.go [hunk 5]
```diff
@@ -1610,39 +1557,7 @@ func resourceServiceRead(ctx context.Context, d *schema.ResourceData, m interfac
...
-			"target": map[string]interface{}{
-				"client_certificate": service.CreateServiceSpec.Spec.Backend.Target.ClientCertificate,
-				"name":               service.CreateServiceSpec.Spec.Backend.Target.Name,
-				"port":               backendPort,
-				"tls":                service.CreateServiceSpec.Spec.Backend.Target.TLS,
-				"tls_insecure":       service.CreateServiceSpec.Spec.Backend.Target.TLSInsecure,
-			},
-		},
-	}
+	spec, diagnostics := flattenServiceSpec(service.CreateServiceSpec.Spec)
```
Comment (dd76) [inline banyan/resource_service.go:1560]: Generally using err is by convention and easy to read instead of diagnostics.

File: banyan/resource_service.go [hunk 6]
```diff
@@ -1659,3 +1574,256 @@ func resourceServiceDelete(ctx context.Context, d *schema.ResourceData, m interf
...
+	v := make(map[string]interface{})
+	v["frontend_address"] = flattenServiceFrontendAddresses(toFlatten.FrontendAddresses)
+	v["host_tag_selector"] = toFlatten.HostTagSelector
+	v["tls_sni"] = toFlatten.TLSSNI
+	flattened = append(flattened, v)
+	return
+}
+
+func flattenServiceFrontendAddresses(toFlatten []service.FrontendAddress) (flattened []interface{}) {
+	flattened = make([]interface{}, len(toFlatten), len(toFlatten))
```
Comment (dd76) [inline banyan/resource_service.go:1600]: Good to use slice and append as in other places

File: banyan/resource_service.go [hunk 7]
```diff
@@ -1659,3 +1574,256 @@ func resourceServiceDelete(ctx context.Context, d *schema.ResourceData, m interf
...
+	v["connector_name"] = toFlatten.ConnectorName
+	v["dns_overrides"] = toFlatten.DNSOverrides
+	v["http_connect"] = toFlatten.HTTPConnect
+	v["backend_allowlist"] = toFlatten.Whitelist
+
+	flattened = append(flattened, v)
+	return
+}
+
+func flattenServiceAllowPatterns(toFlatten []service.BackendAllowPattern) (flattened []interface{}) {
```
Comment (dd76) [inline banyan/resource_service.go:1624]: Same here, we have slice , plz use append() instead of array constructs.

File: banyan/resource_service.go [hunk 8]
```diff
@@ -1659,3 +1574,256 @@ func resourceServiceDelete(ctx context.Context, d *schema.ResourceData, m interf
...
+		flattened[idx] = v
+	}
+	return
+}
+
+func flattenServiceTarget(toFlatten service.Target) (flattened []interface{}, diagnostics diag.Diagnostics) {
+	v := make(map[string]interface{})
+	port, err := strconv.Atoi(toFlatten.Port) // need to convert this to int
+	if err != nil {
+		diagnostics = diag.Errorf("Could not convert BackendTarget.spec.backend.target.port int to string %v", toFlatten.Port)
```
Comment (dd76) [inline banyan/resource_service.go:1661]: There is an error, return err.

File: banyan/resource_service.go (hunk 5 above)
Comment (TheMattSchiller) [reply banyan/resource_service.go:1576]: Terraform has a diag.Diagnostics type which is recommended by HashiCorp for debugging the terraform provider https://learn.hashicorp.com/tutorials/terraform/provider-debug

File: banyan/resource_service.go [hunk 9]
```diff
@@ -837,7 +805,7 @@ func resourceServiceCreate(ctx context.Context, d *schema.ResourceData, m interf
 	for _, item := range spec {
 		ii, ok := item.(map[string]interface{})
 		if !ok {
-			err := errors.New("Couldn't type assert element in metadatatags")
+			err := errors.New("Couldn't type assert element in spec")
```
Comment (yoshiotu) [inline banyan/resource_service.go:811]: Could you get rid of `err` altogether:

```
diagnostics = diagnostic.Errorf("Couldn't type assert element in spec")
return
```

File: banyan/resource_service.go [hunk 10]
```diff
@@ -946,52 +908,43 @@ func resourceServiceCreate(ctx context.Context, d *schema.ResourceData, m interf
...
+					diagnostics = createTypeAssertDiagnostic("port", frontEndAddressItemMap["port"])
 					return
 				}
-				svc.Spec.Attributes.FrontendAddresses = append(svc.Spec.Attributes.FrontendAddresses, service.FrontendAddress{
-					CIDR: cidr,
-					Port: strconv.Itoa(port),
-				})
+				newFrontEndAddress.Port = port
 
+				svc.Spec.Attributes.FrontendAddresses = append(svc.Spec.Attributes.FrontendAddresses, newFrontEndAddress)
```
Comment (yoshiotu) [inline banyan/resource_service.go:937]:
```
svc.Spec.Attributes.FrontendAddresses = append(
        svc.Spec.Attributes.FrontendAddresses,
        service.FrontendAddress {
            CIDR: cidr,
            Port: port,
        },
    )
```

File: banyan/resource_service.go [hunk 11]
```diff
@@ -946,52 +908,43 @@ func resourceServiceCreate(ctx context.Context, d *schema.ResourceData, m interf
...
 			}
-			hostTagSelector, ok := jj["host_tag_selector"].([]interface{})
+
+			hts, ok := jj["host_tag_selector"].([]interface{})
 			if !ok {
-				diagnostics = diag.Errorf("Couldn't type assert host_tag_selector")
+				diagnostics = diag.Errorf("couldn't type assert host_tag_selector with type: %v", reflect.TypeOf(jj["host_tag_selector"]))
```
Comment (yoshiotu) [inline banyan/resource_service.go:942]: Would it work to use `%T` here and avoid the need to call `reflect.TypeOf()`?

File: banyan/resource_service.go [hunk 12]
```diff
@@ -1571,9 +1524,18 @@ func resourceServiceRead(ctx context.Context, d *schema.ResourceData, m interfac
 		return
 	}
 	log.Printf("#### readService: %#v", service)
-	d.Set("name", service.ServiceName)
-	d.Set("description", service.Description)
-	d.Set("cluster", service.ClusterName)
+	err = d.Set("name", service.ServiceName)
+	if err != nil {
+		return nil
```
Comment (yoshiotu) [inline banyan/resource_service.go]: Should the `err` value get logged or something?

File: banyan/resource_service.go (hunk 5 above)
Comment (yoshiotu) [reply banyan/resource_service.go:1584]: Should there be a check to see if `diagnostics` contains a fatal error and return immediately?

File: banyan/resource_service.go [hunk 13]
```diff
@@ -1659,3 +1595,256 @@ func resourceServiceDelete(ctx context.Context, d *schema.ResourceData, m interf
...
+		v := make(map[string]interface{})
+		v["origin_header"] = item.OriginHeader
+		v["target"] = item.Target
+		flattened[idx] = v
+	}
+	return
+}
+
+func flattenServiceHTTPHealthCheck(toFlatten service.HTTPHealthCheck) (flattened []interface{}) {
+	v := make(map[string]interface{})
```
Comment (yoshiotu) [inline banyan/resource_service.go:1755]: Why is `flattened` an array? There's only ever one element.

Also, I think this style looks a little nicer:

```
v := map[string]interface{}{
    "addresses": toFlatten.Addresses,
    "enabled": toFlatten.Enabled,
...
}

flattened = []interface{}{v}
return
```

Comment (yoshiotu) [reply banyan/resource_service.go:1755]: I guess you're sticking to a single pattern for all the flatten functions. Seems ok.

File: banyan/resource_service.go [hunk 14]
```diff
@@ -1659,3 +1605,266 @@ func resourceServiceDelete(ctx context.Context, d *schema.ResourceData, m interf
...
+}
+
+func validatePort() func(val interface{}, key string) (warns []string, errs []error) {
+	return func(val interface{}, key string) (warns []string, errs []error) {
+		v, err := typeSwitchPort(val)
+		if err != nil {
+			errs = append(errs, err)
+			return
+		}
+		if v < 0 || v > math.MaxUint16 {
```
Comment (yoshiotu) [inline banyan/resource_service.go:1828]: Should this be `if v <= 0 || v >= math.MaxUint16`? Corresponds to allowed range `[1..65535]`.

**Review by dd76 — COMMENTED** [review]
Looks good to me. Plz take care of yoshio's comments

**Review by dsrinibanyan — APPROVED** [review]
